File: cmd/kavach/internal/commands/session/end.go
package session

import (
	"fmt"

	"github.com/claude/shared/pkg/enforce"
	"github.com/spf13/cobra"
)

var endCmd = &cobra.Command{
	Use:   "end",
	Short: "End session and save state",
	Long: `[END]
desc: End current session and persist state
hook: Stop
purpose: Save session state before termination

[SAVES]
- Session ID and date
- Research done state
- Memory queried state
- CEO/Aegis invocation states

[USAGE]
kavach session end

[OUTPUT]
[END]   date, session, project
[STATE] research, memory, ceo, aegis states`,
	Run: runEndCmd,
}

func runEndCmd(cmd *cobra.Command, args []string) {
	ctx := enforce.NewContext()
	session := enforce.GetOrCreateSession()

	fmt.Println("[END]")
	fmt.Printf("date: %s\nsession: %s\nproject: %s\n\n", ctx.Today, session.ID, session.Project)

	fmt.Println("[STATE]")
	fmt.Printf("research_done: %s\nmemory: %s\nceo: %s\naegis: %s\n",
		boolStr(session.ResearchDone), boolStr(session.MemoryQueried),
		boolStr(session.CEOInvoked), boolStr(session.AegisVerified))

	session.Save()
}

File: crates/kavach-cli/src/commands/session/end.rs
//! Session end: persist state before termination.
//! Hook: Stop

use std::io::Write;

use crate::session;

pub fn run() -> anyhow::Result<()> {
    let sess = session::get_or_create_session();
    let out = std::io::stdout();
    let mut w = out.lock();

    writeln!(w, "[END]")?;
    writeln!(w, "date: {}", sess.today)?;
    writeln!(w, "session: {}", sess.id)?;
    writeln!(w, "project: {}", sess.project)?;
    writeln!(w)?;

    writeln!(w, "[STATE]")?;
    writeln!(w, "research_done: {}", sess.research_done)?;
    writeln!(w, "memory: {}", sess.memory_queried)?;
    writeln!(w, "ceo: {}", sess.ceo_invoked)?;
    writeln!(w, "aegis: {}", sess.aegis_verified)?;

    let _ = sess.save();

    Ok(())
}
